README.md
# project-lava
Optimized Packet Passing for Virtual Networks

  - mainScript.sh <flags>
    * -t Will run test
    * -w will push test csv files to the git wiki repo
    * currently they need to be run indiviudally.
  
  - testScript.sh
    * no flags this script will run framework n:m up to n = 8, m = 8
    * Called by mainScript.sh
    * Should be a sudo call since we work with threads.


FrameworkMod/ChangeLog.txt
2/21/2019 - Cody

	The main point of this was to separate the input queues from the output queues.
Initially I only did that but then there was some weird logic where threadIDs[],
threadArgs[], locks[], and finished[] were still on the old numbering system
(0 through m+n) while queues[] was on the new numbering system. I fixed this by
creating and intializing structs in global.h that encapsulate all these in an 
input struct and an output struct. So for example if you want to access output
queue 3 you would say output.queue[2] and if you want to lock on that thread
you would use output.locks[2].
	I noticed inputQueueCount and outputQueueCount were getting passed around as 
arguments a lot so I added them to the input/output structs and replaced all 
mentions with input.queueCount and output.queueCount. 
	I updated algorithm.c to use the modified framework and everything seems
to be working. Other than that I tried to change as little as possible.
You can use the same Makefile in this folder to test it out.

Files modified: framework.c, global.h, algorithm.c, algorithm.h

2/22/2019 - Alex
	-Added random number generator for generating the length member of the packet structure. 
		-Min payload size is 64 and max payload size is 9000
			-This translates to min packet struct size of 256 and max packet struct size of 9192

	-Added Comments to global.h file (no code added)

2/23/2019 - Alex
	-detached input and processing threads
	-Explicitly set expected array in process_thread() to all zeros
	-Included <string.h> in global.h file in order to use memset()
	-Included <math.h> in order to use in algorithm Files
	-Modified (added -lm) Makefile to work with <math.h> functions 

	Files modified: framework.c, global.h, Makefile

2/23/2019 - Regi
	- Implemented function to change the seed of rand(), so that our results aren't skewed towards a single data set
	- changed minimum payload size to 0, since there are packets that just transfer the header with no payload.
		- minimum packet struct size is now 24 and max is 9024 (in terms of bytes)
	- set expected[], line 96 in framework.c, to the order of the first packet it recieved 
	- moved original algorithm.c (singe wire algorithm) to newly created algorithm folder and renamed it to singleWire.c
	- detached main algorithm thread
	- defined HEADER_SIZE 24, line 36 in global.h. Defines first 24 bytes in packet header
	- changed length variable of packet to correspond to payload size instead of header size
	- implemented some error handling in the framework
	- changed globally defined variable to be represented in bytes. There was an issue in regards to calculating the size of the structures and the size
		in terms of bits and bytes were not consistent. I chose bytes since that's what most functions will use when calculating size.
	- tried testing out memcpy() for the current algorithm, but there seems to be issues possibly because of thread premption or memory ordering

	Files modified: framework.c, global.h
	
2/24/2019 - Lito
	- changed this line $(CC) $(LDFLAGS) -c -o $@ $< $(CFLAGS)
	  $(CC) $(LDFLAGS) -std=c99 -c -o $@ $< $(CFLAGS)
	  So that the for loops don't have issues building on linux
	- changed asm to __asm__ because this is required for c99 as well.
	Files Modified: Makefile, global.h, testScript.sh
	
	


FrameworkMod/global.h
#ifndef GLOBAL_H
#define GLOBAL_H

#define _GNU_SOURCE
#include <stdio.h>
#include <stdint.h>
#include <unistd.h>
#include <stdlib.h>
#include <assert.h>
#include <errno.h>
#include <sched.h>
#include <string.h>
#include <math.h>
#include <time.h>

//Constants for upper limit of queues
#define MAX_NUM_QUEUES 16
#define MAX_NUM_INPUT_QUEUES 8
#define MAX_NUM_OUTPUT_QUEUES 8

//Constants for threads
#define MAX_NUM_THREADS 20
#define BUFFERSIZE 256

//How many packets need to be processed by each queue before the program terminates
//We could also make this time instead
#define RUNTIME 50000

//Used for calibrating the amount of clock cycles in a second
#define CAL_DUR 2000000ULL
#define CALIBRATION 5

#define MIN_PACKET_SIZE 24
#define MAX_PACKET_SIZE 9024

#define MAX_PAYLOAD_SIZE 9000
#define FLOWS_PER_QUEUE 8

#define HEADER_SIZE 24

#define FENCE() \
   __asm__ volatile ("mfence" ::: "memory");

typedef unsigned long long tsc_t;

//Data structure to represent a packet.
//length (size_t) - The total size of the packet
//flow (size_t) - The flow of the packet
//order (size_t) - The order of the packet within its flow
//data (size_t) - Payload of the packet
typedef struct packet{ 
    size_t length;
    size_t flow;
    size_t order;
    unsigned char data[MAX_PAYLOAD_SIZE];
}packet_t;

//Data structure for Queue:
//toRead (size_t) - The next unread position in the queue
//toWrite (size_t) - The next position to write to in the queue
//count (size_t) - The number of packets read by the queue
//data (packet_t array - Space for packets in the queue
typedef struct Queue {
    size_t toRead;
    size_t toWrite;
    size_t count;
    packet_t data[BUFFERSIZE];
}queue_t;

//Arguments to be passed to input threads
//firstQueue (queue_t) - pointer to the first queue for the thread to write to/process
//firstQueueNum (size_t) - The index of the queue in the array of queues
typedef struct threadArgs{
    queue_t *queue;
    size_t queueNum;
}threadArgs_t;

//input and output struct are mostly identical right now and mainly used to
//abstract the inputs resources from the outputs resources but we may want
//to add unique characteristics to each at some point.

//Initialize items for input threads
//queueCount (size_t) - The number of input queues
//queues (queue_t) - The input queue struct
//queues (queue_t) - The input queue struct
//threadIds (pthread_t) - The Id's for ech input thread
//threadArgs (threadArgs_t) - Arguments to be passed to input threads
//locks (pthread_mutex_t) - lock for each input queue
//finished (size_t) - flag that indicates when a given thread has generated x number of packets
typedef struct input{
    size_t queueCount;
    pthread_t threadIDs[MAX_NUM_INPUT_QUEUES]; 
    threadArgs_t threadArgs[MAX_NUM_INPUT_QUEUES];
    pthread_mutex_t locks[MAX_NUM_INPUT_QUEUES];
    queue_t queues[MAX_NUM_INPUT_QUEUES];
    size_t finished[MAX_NUM_INPUT_QUEUES];
}input_t;
input_t input;

//Initialize items for output threads
//queueCount (size_t) - The number of output queues
//queues (queue_t) - The output queue struct
//threadIds (pthread_t) - The Id's for ech output thread
//threadArgs (threadArgs_t) - Arguments to be passed to output threads
//locks (pthread_mutex_t) - lock for each output queue
//finished (size_t) - flag that indicates when a given thread has processed x number of packets
typedef struct output{
    size_t queueCount;
    pthread_t threadIDs[MAX_NUM_OUTPUT_QUEUES]; 
    threadArgs_t threadArgs[MAX_NUM_OUTPUT_QUEUES];
    pthread_mutex_t locks[MAX_NUM_OUTPUT_QUEUES];
    queue_t queues[MAX_NUM_OUTPUT_QUEUES];
    size_t finished[MAX_NUM_OUTPUT_QUEUES];
}output_t;
output_t output;

//Used to record roughly the number of ticks in a second
tsc_t clockSpeed;

static __inline__ tsc_t rdtsc(void);
void printLoading(int);
tsc_t cal(void); 

#endif
